vllm/worker/model_runner.py

```diff
@@ -845,13 +921,67 @@ def execute_model(
         if not sampling_metadata.perform_sampling:
             return None
 
+        if self.is_mamba:
+            for i, offset in enumerate(indices):
+                self.mamba_cache[0][:, offset].copy_(conv_state[:, i])
+                self.mamba_cache[1][:, offset].copy_(ssm_state[:, i])
+
         # Sample the next token.
         output = self.model.sample(
             logits=logits,
             sampling_metadata=sampling_metadata,
         )
         return output
 
+    def _get_first_free_mamba_cache_index(self):
+        max_possible_bs = self.mamba_cache[0].shape[1]
+        occupied = [
+            id for seq_ids in self.request2i.values()
+            for id in seq_ids.values()
+        ]
+        first_free_index = [i not in occupied
+                            for i in range(max_possible_bs)].index(True)
+        return first_free_index
+
+    def _prepare_request_mamba_cache(self, requests_info: List[RequestInfo],
+                                     batch_size: int):
+        indices = []
+        max_possible_bs = self.mamba_cache[0].shape[1]
+        for request_info in requests_info:
+            cur_rid = request_info.request_id
+            if cur_rid not in self.request2i:
+                self.request2i[cur_rid] = {}
+                for seq_id in request_info.seqs_id:
+                    f_free_index = self._get_first_free_mamba_cache_index()
+                    self.request2i[cur_rid][seq_id] = f_free_index
+                    indices.append(f_free_index)
+            else:
+                for seq_id in request_info.seqs_id:
+                    if seq_id not in self.request2i[cur_rid]:
+                        f_free_index = self._get_first_free_mamba_cache_index()
+                        ## case of decoding n>1
+                        i_exist = list(self.request2i[cur_rid].values())[0]
+                        self.mamba_cache[0][:, f_free_index].copy_(
+                            self.mamba_cache[0][:, i_exist])
+                        self.mamba_cache[1][:, f_free_index].copy_(
+                            self.mamba_cache[1][:, i_exist])
+                        self.request2i[cur_rid][seq_id] = f_free_index
+                    indices.append(self.request2i[cur_rid][seq_id])
```

Comment:
In vLLM, we typically put the logic of figuring out the memory address for each request in the driver instead of worker/model runner. For Transformer-based models, the logic is implemented in `BlockSpaceManager` class. I believe we should do the same thing for mamba as well. We can create another `FixedSpaceManager` class to manage the memory for mamba-style models.